### ISSUE TITLE
Eglretrace doesn't work with software rendering

### ISSUE THREAD

**Issue by sstewartgallus** [apitrace/apitrace#313]
Capturing the output of any EGL using program (even `es2gears`) succeeds and the trace can be rerun by `eglretrace` but not when using software rendering (such as with the `LIBGL_ALWAYS_SOFTWARE=true` option). It seems to fail with a BadMatch parameter error (even if the trace was recorded using software rendering originally). Testing using software rendering would be useful for sanity checking purposes.
~~~
$ es2_info
libEGL warning: failed to create a pipe screen for i965
EGL_VERSION: 1.4 (DRI2)
EGL_VENDOR: Mesa Project
EGL_EXTENSIONS:
    EGL_MESA_drm_image, EGL_WL_bind_wayland_display, EGL_KHR_image_base, 
    EGL_KHR_image_pixmap, EGL_KHR_image, EGL_KHR_gl_texture_2D_image, 
    EGL_KHR_gl_texture_cubemap_image, EGL_KHR_gl_renderbuffer_image, 
    EGL_KHR_surfaceless_context, EGL_KHR_create_context, EGL_NOK_swap_region, 
    EGL_NOK_texture_from_pixmap, EGL_EXT_create_context_robustness, 
    EGL_EXT_image_dma_buf_import, EGL_NV_post_sub_buffer
EGL_CLIENT_APIS: OpenGL OpenGL_ES OpenGL_ES2 OpenGL_ES3 
GL_VERSION: OpenGL ES 3.0 Mesa 10.1.3
GL_RENDERER: Mesa DRI Intel(R) Bay Trail 
GL_EXTENSIONS:
    GL_EXT_blend_minmax, GL_EXT_multi_draw_arrays, 
    GL_EXT_texture_filter_anisotropic, GL_EXT_texture_compression_dxt1, 
    GL_EXT_texture_format_BGRA8888, GL_OES_compressed_ETC1_RGB8_texture, 
    GL_OES_depth24, GL_OES_element_index_uint, GL_OES_fbo_render_mipmap, 
    GL_OES_mapbuffer, GL_OES_rgb8_rgba8, GL_OES_standard_derivatives, 
    GL_OES_stencil8, GL_OES_texture_3D, GL_OES_texture_npot, 
    GL_OES_EGL_image, GL_OES_depth_texture, GL_OES_packed_depth_stencil, 
    GL_EXT_texture_type_2_10_10_10_REV, GL_OES_get_program_binary, 
    GL_APPLE_texture_max_level, GL_EXT_discard_framebuffer, 
    GL_EXT_read_format_bgra, GL_NV_fbo_color_attachments, 
    GL_OES_EGL_image_external, GL_OES_vertex_array_object, 
    GL_ANGLE_texture_compression_dxt3, GL_ANGLE_texture_compression_dxt5, 
    GL_EXT_texture_rg, GL_EXT_unpack_subimage, GL_NV_draw_buffers, 
    GL_NV_read_buffer, GL_EXT_map_buffer_range, GL_OES_depth_texture_cube_map, 
    GL_OES_surfaceless_context, GL_EXT_color_buffer_float, 
    GL_EXT_shader_integer_mix
~~~

**Comment by jrfonseca** [comment]
I have a similar setup, and indeed when `LIBGL_ALWAYS_SOFTWARE=true` causes `eglSwaBuffers` to trigger

    error: xlib: BadMatch (invalid parameter attributes)

X errors.

Why it's not clear.

In fact, I get this error even without `eglretrace`. For example, the sample apps I have in https://github.com/apitrace/apitrace-tests/tree/master/apps/egl/gles2 fail simialrly, even though they are using GLFW -- a totally distinct code base from Apitrace which fails the same way.

But indeed the Mesa demos like `es2_gears` and `es2_tri`  from http://cgit.freedesktop.org/mesa/demos/tree/src/egl/opengles2 don't have this problem somehow.

I'm inclined to suspect a bug in DRI software rasterizers -- it's not the first time I come a across weird behavior in the DRI software rasterizer due to slight mismatches between the X visuals.

But I'm not sure.  I already spend sometime digging into this, but I can't afford to put much more time.

**Comment by jrfonseca** [comment]
Yep, my hunch was right. Mesa EGL implementation is busted when `LIBGL_ALWAYS_SOFTWARE=true`.

The problem is that, if we request an EGL config with a non-zero `EGL_ALPHA_SIZE` then in http://cgit.freedesktop.org/mesa/mesa/tree/src/egl/drivers/dri2/platform_x11.c#n264 

    swrastCreateDrawable(dri2_dpy, dri2_surf, _eglGetConfigKey(conf, EGL_BUFFER_SIZE));

the `EGL_BUFFER_SIZE` will return a depth of 32bits, which is later passed to `xcb_put_image` causing X11 to return an error, because it expects a 24bits depth.

This change works around it:

    $ git diff retrace/glws_egl_xlib.cpp | github-quote 
    diff --git a/retrace/glws_egl_xlib.cpp b/retrace/glws_egl_xlib.cpp
    index 81058d6..bae02cc 100644
    --- a/retrace/glws_egl_xlib.cpp
    +++ b/retrace/glws_egl_xlib.cpp
    @@ -311,5 +311,5 @@ createVisual(bool doubleBuffer, unsigned samples, Profile profile) {
         attribs.add(EGL_GREEN_SIZE, 1);
         attribs.add(EGL_BLUE_SIZE, 1);
    -    attribs.add(EGL_ALPHA_SIZE, 1);
    +    attribs.add(EGL_ALPHA_SIZE, 0);
         attribs.add(EGL_DEPTH_SIZE, 1);
         attribs.add(EGL_STENCIL_SIZE, 1);

eglretrace doesn't use exactly the same visual. It could sometimes, but not always.  Alpha channel is sometimes necessary for blending, so it's a better value than no alpha.

In short, not our bug. You'll need to file a Mesa bug on bugs.freedesktop.org for this.  I'm also a Mesa developer but I don't maintain that code.

**Comment by jrfonseca** [comment]
Actually, I've gotten to the bottom of this, and posted a patch that fixes it to http://lists.freedesktop.org/archives/mesa-dev/2015-January/074693.html

**Comment by sstewartgallus** [comment]
Thank you very much! I use EGL and X11 very directly and without any abstracted libraries so I guess that's why I didn't get the error with software rendering.

> eglretrace doesn't use exactly the same visual. It could sometimes, but not always. Alpha channel is sometimes necessary for blending, so it's a better value than no alpha.

Personally, I think it should use as close a visual as possible but I'm guessing you have a good reason for that? Or is it just that that is a feature that has not been implemented yet and nobody has requested it yet and you have more important work to do anyways (which is fine)?

**Comment by jrfonseca** [comment]
> Personally, I think it should use as close a visual as possible but I'm guessing you have a good reason for that? Or is it just that that is a feature that has not been implemented yet and nobody has requested it yet and you have more important work to do anyways (which is fine)?

Yes, precisely.

This is briefly documented in https://github.com/apitrace/apitrace/blob/master/docs/BUGS.markdown#retracing , but the full rationale is:

* it's more important to be able to move traces between different platforms than to replay EGL/WGL/GLX/etc API calls verbatim; so instead of replaying these calls as-is, we have an abstraction layer that emulates these calls on top of whatever the host API has

* extracting the visual information is not impossible but it's a fair amount of work to reverse engineer the visual attributes the application used in the end, specially when there are so many ways an application can choose a visual (e.g., it can ask the EGL/WGL/GLX to choose a visual that best matches certain attributes, or it can dump all supported attributes and choose one internally based on any criteria it wants),

* 99% of the apps just want a RGB8 or RGBA8 visual, with a depth and stencil buffer

We do however go to extra length to reverse engineer the GL version and profile that the app specified
